On-Chain Cash Exchange (OCEX) Milestone 1

dsrdrk11a opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [x] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [x] This pull request is being made by the same account as the accepted application.
- [x] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [x] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/803.

semuelle commented:
Thank you for the delivery, @dsrdrk11a. Is there a reason why you left out deliverables 0d and 0e from this delivery?

We will look into it as soon as possible. Please excuse us if this might take longer than usual, we are currently experiencing a bit of a backlog.

dsrdrk11a commented:
@semuelle , yeah, we significantly corrected all the milestones while applying for this grant and did not check the compliance of these fields with the scope of work that was finally left in the milestones.
In Milestone 1 we don't deliver any executable files so there's no need in the docker file (0d line) as we see it. But we will provide a docker file in milestone 2. As for the 0e (article) - firstly it was only in the 3rd Milestone and when it was deleted we mistakenly added it to all left milestones, but we planned to write an overall article on the whole project. So if it is ok we'll do it in Milestone 2. But if it is mandatory now, we'll think how to split our article in two parts.

dsrdrk11a commented:
@semuelle , @takahser 
thank you for your comments. [Amendment](https://github.com/w3f/Grants-Program/pull/907)

takahser reviewed (changes requested):
@dsrdrk11a please accept my apologies for the delay here and let me thanks for your patience. 🙏

I saw that currently you're missing the LICENSE file. You linked the [Apache 2.0 License](https://choosealicense.com/licenses/apache-2.0/), where it says:

> Create a text file (typically named LICENSE or LICENSE.txt) in the root of your source code and copy the text of the license into the file.

Please follow these steps and add a file with the appropriate naming and contents.

Also, I took a first review and summarised them in my evaluation which you can find here: https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/ocex_1_takahser.md

Lmk if there is anything unclear or if you have any further questions.

dsrdrk11a commented:
@takahser , thank you for your feedback!
I've corrected the license link and added the 3.2 deliverable with a link to the coupon validation method (as we've put it in one deliverable with the registration method).
Regarding the `transfer_ownership` method test - we've put it [here](https://github.com/bsn-si/ocex-smartcontract/blob/main/lib.rs#L445)

Also I want to comment about the build fail.
The build is tested on nightly-2022-03-14-x86_64-unknown-linux-gnu toolchain with rustc 1.62.0-nightly (e85edd9a8 2022-04-28) version. If you have compilation errors, please try changing the toolkit and compiler version to the specified one. We've also added this comment to the [Readme](https://github.com/bsn-si/ocex-smartcontract/blob/f9d3dc4ec01c713f35df9e4af410c722b9344f64/README.md#resolve-common-errors)

takahser reviewed (changes requested):
@dsrdrk11a sorry for the wait, I'm still behind schedule. I tried to update the Rust version to yours and also to use the rust toolchain you're using, but the build still fails. I updated my evaluation and included more details there: https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/ocex_1_takahser.md

Feel free to reach out if you have further questions.

dsrdrk11a commented:
Hello, @takahser
The warning message is fixed.

About the Rust version: the `rustc 1.63.0-nightly (4c5f6e627 2022-05-17)` might also work, please try. But according to your case it seems like the Rust version is not a problem here, you have an error during the wasm optimization step. Please, make sure you have [binaryen](https://github.com/WebAssembly/binaryen) installed. That might fix it.

We've also corrected the [Readme](https://github.com/bsn-si/ocex-smartcontract#resolve-common-errors) accordingly.

takahser reviewed (changes requested):
@dsrdrk11a thanks for your patience and sorry for the delay here.

> The warning message is fixed.

I confirm that the warning in the tests is fixed, I updated that in the [updated evaluation](https://github.com/w3f/Grant-Milestone-Delivery/blob/241c6a80d72553aa5ca21263b1c2db2fa1588d48/evaluations/ocex_1_takahser.md) correspondingly.

> Please, make sure you have [binaryen](https://github.com/WebAssembly/binaryen) installed.

I double-checked, it's actually already installed:

```bash
$ sudo apt-get -y install binaryen
Reading package lists... Done
Building dependency tree       
Reading state information... Done
binaryen is already the newest version (91-1).
```

However, the build still fails with the same error. I also followed your steps outlined in the README, which didn't help neither. For more details see my [updated evaluation](https://github.com/w3f/Grant-Milestone-Delivery/blob/241c6a80d72553aa5ca21263b1c2db2fa1588d48/evaluations/ocex_1_takahser.md).

dsrdrk11a commented:
Hey, @takahser! Great to hear from you!
I'm sorry, that it didn't work again. Could you, please, provide more info about the problem to us.
Please, check your `wasm-opt` version with the `wasm-opt --version` command, run compilation with `RUST_BACKTRACE=full cargo +nightly contract build` option and send the full log of the `cargo` compiler. We'll figure out the reason.

dsrdrk11a commented:
@takahser, thank you for more info.
Trying to find out the reasons, we’ve executed install and build of the smartcontract in different environments like MacOS and Linux including Ubuntu OS on a clean virtual machine and everything worked well. Here’s the [video log](https://github.com/bsn-si/ocex-smartcontract/blob/main/README.md#video-guide)
We’ve executed the following:
```
sudo apt install build-essential pkg-config libssl-dev curl git binaryen
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs/ | sh # nightly version complete bundle
source $HOME/.cargo/env
cargo install cargo-dylint dylint-link cargo-contract
git clone https://github.com/bsn-si/ocex-smartcontract
cd ocex-smartcontract/
cargo contract build
```
So it’s difficult to identify the reason why it fails on your side without a detailed log.

To overcome this, we’ve made a docker set-up so you can build the smartcontract in a container, isolated from the main environment. Please, follow the steps, described [here](https://github.com/bsn-si/ocex-smartcontract#building-in-docker). We’ve successfully tested it in several environments from MacOS to Linux.

takahser commented:
Hey @dsrdrk11a 

Sorry for the wait, I was at a company event during the whole last week and am in the process of working off my backlog.

I ran the commands you shared. However, it failed at step 4:

```bash
$ cargo install cargo-dylint dylint-link cargo-contract
    Updating crates.io index
  Downloaded cargo-dylint v2.0.6
  Downloaded 1 crate (23.4 KB) in 0.17s

  (...)

    Compiling dylint_internal v2.0.6
error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1185:76
     |
1185 |                     format!("failed to parse value from --config argument `{arg}` as a dotted key expression")
     |                                                                            ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1207:61
     |
1207 | ...                   "--config argument `{arg}` \
     |                                           ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1215:57
     |
1215 | ...                   "--config argument `{arg}` \
     |                                           ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1222:61
     |
1222 | ...                   "--config argument `{arg}` \
     |                                           ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1231:57
     |
1231 | ...                   "--config argument `{arg}` \
     |                                           ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1237:59
     |
1237 | ...                   bail!("--config argument `{arg}` doesn't provide a value")
     |                                                 ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1245:45
     |
1245 |                         "--config argument `{arg}` was not a TOML dotted key expression (such as `build.jobs = 2`)"
     |                                             ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1250:76
     |
1250 |                     format!("failed to parse value from --config argument `{arg}`")
     |                                                                            ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1254:84
     |
1254 |                     .with_context(|| format!("failed to convert --config argument `{arg}`"))?
     |                                                                                    ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `arg`
    --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/config/mod.rs:1262:78
     |
1262 |                 .with_context(|| format!("failed to merge --config argument `{arg}`"))?;
     |                                                                              ^^^^^
     |
     = help: if you intended to capture `arg` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `name`
   --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/toml/mod.rs:491:48
    |
491 |             profile.validate_profile(&format!("{name}.build-override"), features)?;
    |                                                ^^^^^^
    |
    = help: if you intended to capture `name` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `name`
   --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/toml/mod.rs:497:52
    |
497 |                 profile.validate_profile(&format!("{name}.package.{override_name}"), features)?;
    |                                                    ^^^^^^
    |
    = help: if you intended to capture `name` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: there is no argument named `override_name`
   --> /home/xxx/.cargo/registry/src/github.com-1ecc6299db9ec823/cargo-0.61.0/src/cargo/util/toml/mod.rs:497:67
    |
497 |                 profile.validate_profile(&format!("{name}.package.{override_name}"), features)?;
    |                                                                   ^^^^^^^^^^^^^^^
    |
    = help: if you intended to capture `override_name` from the surrounding scope, add `#![feature(format_args_capture)]` to the crate attributes

error: could not compile `cargo` due to 13 previous errors
warning: build failed, waiting for other jobs to finish...
  Installing dylint-link v2.0.6
   Compiling proc-macro2 v1.0.39

  (...)

  Downloaded secp256k1-sys v0.5.2
  Downloaded 82 crates (5.9 MB) in 0.49s (largest was `secp256k1-sys` at 2.5 MB)
error: failed to compile `cargo-dylint v2.0.6`, intermediate artifacts can be found at `/tmp/cargo-install2duMDM`

Caused by:
  build failed
error: failed to compile `cargo-contract v1.4.0`, intermediate artifacts can be found at `/tmp/cargo-install1yJWDl`

Caused by:
  package `async-global-executor v2.1.0` cannot be built because it requires rustc 1.59 or newer, while the currently active rustc version is 1.58.0-nightly
     Summary Successfully installed dylint-link! Failed to install cargo-dylint, cargo-contract (see error(s) above).
error: some crates failed to install
```

However, surprisingly the contract is able to be built, if the nightly param is skipped:

```bash
$ cargo contract build
 [1/5] Building cargo project
    Updating crates.io index
    Updating git repository `https://github.com/paritytech/substrate.git`
  Downloaded compiler_builtins v0.1.49
  Downloaded hashbrown v0.11.0

    Finished release [optimized] target(s) in 13.19s
     Running `target/ink/release/metadata-gen ''`
 [5/5] Generating bundle

Original wasm size: 118.3K, Optimized: 78.6K

The contract was built in DEBUG mode.

Your contract artifacts are ready. You can find them in:
/home/xxx/repos/ocex-smartcontract/target/ink

  - ocex.contract (code + metadata)
  - ocex.wasm (the contract's code)
  - metadata.json (the contract's metadata)
```

In contrary to that, the build using nightly still fails:

```bash
$ cargo +nightly contract build
 [1/5] Building cargo project
    Updating crates.io index
    Updating git repository `https://github.com/paritytech/substrate.git`
   Compiling cc v1.0.69
   Compiling memchr v2.4.1
   Compiling proc-macro2 v1.0.39

   (...)

   Compiling ocex v0.1.0 (/tmp/cargo-contract_pcNGXg)
    Finished release [optimized] target(s) in 10.09s
 [2/5] Post processing wasm file
 [3/5] Optimizing wasm file
ERROR: The wasm-opt optimization failed.

The error which wasm-opt returned was: 
```

Finally, I ran the [build using docker](https://github.com/bsn-si/ocex-smartcontract#building-in-docker) (as suggested by you):

```bash
$ docker build -t ocex-contract .
        Sending build context to Docker daemon  309.8kB
    Step 1/4 : FROM paritytech/contracts-ci-linux:latest
    latest: Pulling from paritytech/contracts-ci-linux
    d7bfe07ed847: Pull complete 
    2e43357312c2: Pull complete 
    24bd3e0f3bc5: Pull complete 
    Digest: sha256:3a603c9a596fef5d2fa05d728f4220c26082dfac3e4d9a005eb5535f371c4b19
    Status: Downloaded newer image for paritytech/contracts-ci-linux:latest
    ---> c29c318da494
    Step 2/4 : WORKDIR /contract
    ---> Running in 7711f5658c1b
    Removing intermediate container 7711f5658c1b
    ---> 603b652c5409
    Step 3/4 : COPY . .
    ---> 004a8bc00d0e
    Step 4/4 : RUN cargo contract build --release
    ---> Running in 741179a1bda5
    [1/5] Checking ink! linting rules
    Building driver for toolchain `nightly-2022-03-14-x86_64-unknown-linux-gnu`
        Updating crates.io index
    Downloading crates ...
    (...)
    Downloaded camino v1.0.9
    Compiling proc-macro2 v1.0.39
    (...)
    Compiling dylint_driver-nightly-2022-03-14-x86_64-unknown-linux-gnu v0.1.0 (/tmp/.tmpIqnmoN)
        Finished dev [unoptimized + debuginfo] target(s) in 9.76s
    Checking with toolchain `nightly-2022-03-14-x86_64-unknown-linux-gnu`
    (...)
    Downloaded rustc-demangle v0.1.21
    Compiling compiler_builtins v0.1.71
    Compiling core v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core)
    Compiling libc v0.2.121
    Compiling cc v1.0.69
    Compiling dlmalloc v0.2.3
    Compiling std v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std)
    (...)
    Compiling unwind v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/unwind)
    Compiling quote v1.0.18
    (...)
    Compiling ink_primitives v3.2.0
    Compiling rustc-std-workspace-core v1.99.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/rustc-std-workspace-core)
    Compiling ink_lang_macro v3.2.0
    Compiling alloc v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc)
    Compiling cfg-if v0.1.10
    Compiling rustc-demangle v0.1.21
    Compiling rustc-std-workspace-alloc v1.99.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/rustc-std-workspace-alloc)
    Compiling panic_abort v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/panic_abort)
    Compiling panic_unwind v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/panic_unwind)
    Compiling hashbrown v0.12.0
    Compiling std_detect v0.1.5 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/stdarch/crates/std_detect)
    Compiling proc_macro v0.0.0 (/usr/local/rustup/toolchains/nightly-2022-05-05-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/proc_macro)
    Compiling arrayref v0.3.6
    (...)
    Compiling ocex v0.1.0 (/tmp/cargo-contract_zieJEi)
        Finished release [optimized] target(s) in 22.16s
    [3/5] Post processing wasm file
    [4/5] Optimizing wasm file
    [4/5] Generating metadata
        Updating crates.io index
        Updating git repository `https://github.com/paritytech/substrate.git`
    Downloading crates ...
    (...)
    Downloaded secp256k1-sys v0.5.2
    Compiling proc-macro2 v1.0.39
    (...)
    Compiling ocex v0.1.0 (/tmp/cargo-contract_xJURiB)
    Compiling metadata-gen v0.1.0 (/tmp/cargo-contract_xJURiB/.ink/metadata_gen)
        Finished release [optimized] target(s) in 17.87s
        Running `target/ink/release/metadata-gen ''`
    [5/5] Generating bundle

    Original wasm size: 76.3K, Optimized: 45.0K

    The contract was built in RELEASE mode.

    Your contract artifacts are ready. You can find them in:
    /contract/target/ink

    - ocex.contract (code + metadata)
    - ocex.wasm (the contract's code)
    - metadata.json (the contract's metadata)
    Removing intermediate container 741179a1bda5
    ---> c55699d5ca25
    Successfully built c55699d5ca25
    Successfully tagged ocex-contract:latest
    
    $ docker create -ti --name ocex ocex-contract bash # create image
    77e29af8272af8a1af071a77b9c24eaa3cd0cf8f8d7eefd9af735b65da52b185
    
    $ docker cp ocex:/contract/target/ink/ocex.contract ~/ocex.contract

    $ docker image ls
    REPOSITORY      TAG         IMAGE ID        CREATED         SIZE
    ocex-contract   latest      c55699d5ca25    7 minutes ago   6.54GB
```

So I'm not clear, why `cargo +nightly contract build` still doesn't work. Any idea?

takahser reviewed (changes requested):
See my previous comment.

dsrdrk11a commented:
Hello, @takahser.
`+nightly` flag downloads the latest version of the toolchain, that is updated daily. We can not discourse for sure why the last nightly version fails on your side, as nightly daily versions are not stable by default – if it is working correctly on our side, that doesn’t guarantee, that it will be ok in your particular environment. This is not smt that depends on our code. So when you skip the `+nightly` flag it uses an earlier version of the nightly toolchain that is set up by default (according to your versions list that you have provided) and functions correctly in your environment.
In case of the docker file, the contract is built isolated from your environment.

So as you succeeded in builing the contract correctly I propose to proceed this way.

takahser commented:
@dsrdrk11a thanks for elaborating, that works for me. 👍
In general, everything compiles and builds properly. Please bear with me as I'm reviewing the contracts, you can expect clearer feedback by tomorrow.

takahser reviewed (approved):
@dsrdrk11a thanks for your patience, I can happily confirm that everything works and that I've accepted this delivery. You can find the evaluation [here](https://github.com/w3f/Grant-Milestone-Delivery/blob/64be42dede5eca16e2c8ddb7a85a697ce1bf61f2/evaluations/ocex_1_takahser.md).
